client/state/experiments/reducer.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * External Dependencies
+ */
+import { Action } from 'redux';
+
+/**
+ * Internal Dependencies
+ */
+import { EXPERIMENT_FETCH, EXPERIMENT_ASSIGN, CURRENT_USER_RECEIVE } from 'state/action-types';
+import { ExperimentState, ExperimentAssign } from 'state/experiments/types';
+import { tracksAnonymousUserId } from 'lib/analytics/ad-tracking';
+
+/**
+ * Attempt to get the anon id for the user, if set
+ */
+export function getAnonIdFromCookie(): string | null {
+	const id = tracksAnonymousUserId();
+
+	return id == null || id === '' ? null : id;
+}
+
+const appStartedAt = Date.now();
+
+const resetState: ( anonId: string | null ) => ExperimentState = anonId => ( {
+	anonId,
+	isLoading: true,
+	nextRefresh: appStartedAt,
+	variations: null,
+} );
+
+export default function reducer( state: ExperimentState = resetState( null ), action: Action ) {
+	switch ( action.type ) {
+		/**
+		 * Store the anon id when we first load the application
+		 */
+		case '@@INIT':
+			state.anonId = getAnonIdFromCookie();
```

Comment:
Checking for `@@INIT` action type is not a reliable way to set the initial state value. The type of the initial action can change at any time, and Redux in development mode will randomize it to discourage users from using this pattern.

You need to assign the initial state as a default value of the `state` parameter, like this:
```js
function reducer( state = resetState( getAnonIdFromCookie() ), action ) {
  switch ( action.type ) {
    case EXPERIMENT_ASSIGN:
      // all the cases where we want to return modified state
    default:
      return state; // return unmodified state, and the default value in case it was `undefined`
  }
}
```

Reply:
Thanks! That's actually pretty smart :) I haven't used js in awhile, so I forgot that you can have non-static parameters!

Reply:
Oh, that won't actually work, because it causes every component test to fail with "document not defined." `@@INIT` is probably the best to hook into without causing cascading failures on unrelated tests.

Reply:
`@@INIT` won't work, as such an action is never dispatched at runtime. It's an action that the Redux DevTools plugin dispatches to your store.

If you want to make tests not fail on DOM references, you need to specify Jest environment:
```js
/**
 * @jest-environment jsdom
 */
```
The reducer should be also safe to run in server context, i.e., in the Node.js Calypso server. Then your code needs to check for `typeof document !== 'undefined'` and guard the cookie access.

Reply:
`typeof` was the missing magic. (I was trying `document === undefined`)

Reply:
Yes, just `document !== undefined` fails because the identifier `document` is not defined. `typeof` can accept undefined identifiers and doesn't fail.